File: go/core/context.go
package core

import "context"

type contextKey struct{}

// ContextData holds customer/project attribution independent of tasks.
// It enables dynamic attribution without requiring an explicit task to be started.
type ContextData struct {
	CustomerID string
	ProjectID  string
	Metadata   map[string]interface{}
	Agent      string // used as task_type for auto-created session tasks
}

type trackedTaskKeyType struct{}

var trackedTaskKey = trackedTaskKeyType{}

type contextDataKeyType struct{}

var contextDataKey = contextDataKeyType{}

// SetContext attaches customer attribution data to the context.
// This allows the HTTP adapter and other components to create auto-tasks
// with the correct attribution without an explicit task being started.
func SetContext(ctx context.Context, cd *ContextData) context.Context {
	return context.WithValue(ctx, contextDataKey, cd)
}

// GetContextData returns the ContextData from the context, or nil if not set.
func GetContextData(ctx context.Context) *ContextData {
	v := ctx.Value(contextDataKey)
	if v == nil {
		return nil
	}
	return v.(*ContextData)
}

// ClearContext removes the ContextData from the context.
func ClearContext(ctx context.Context) context.Context {
	return context.WithValue(ctx, contextDataKey, nil)
}

// WithTask returns a new context with the given Task attached.
func WithTask(ctx context.Context, task *Task) context.Context {
	return context.WithValue(ctx, contextKey{}, task)
}

// GetCurrentTask returns the Task from the context, or nil.
func GetCurrentTask(ctx context.Context) *Task {
	v := ctx.Value(contextKey{})
	if v == nil {
		return nil
	}
	return v.(*Task)
}

// WithTrackedTask returns a new context with the given TrackedTask attached.
func WithTrackedTask(ctx context.Context, tt *TrackedTask) context.Context {
	return context.WithValue(ctx, trackedTaskKey, tt)
}

// GetCurrentTrackedTask returns the TrackedTask from the context, or nil.
func GetCurrentTrackedTask(ctx context.Context) *TrackedTask {
	v := ctx.Value(trackedTaskKey)
	if v == nil {
		return nil
	}
	return v.(*TrackedTask)
}

// LinkParent sets child.ParentTaskID from the context's current task,
// but only if ParentTaskID is not already set.
func LinkParent(ctx context.Context, child *Task) {
	if child.ParentTaskID != nil {
		return
	}
	parent := GetCurrentTask(ctx)
	if parent != nil {
		id := parent.TaskID
		child.ParentTaskID = &id
	}
}

// ---------------------------------------------------------------------------
// Per-call network-event suppression flag (v1 §5.3 invariant)
// ---------------------------------------------------------------------------
//
// When set on a context, the HTTP adapter still records bytes into the
// NetworkAccountant but does NOT emit a standalone `network` event. LLM
// instruments wrap their outbound HTTP call with WithSuppressNetworkEvent
// so each call produces at most one of {llm_call, external_cost, network}
// — the v1 §5.3 "≤ 1 event per HTTP call" invariant.

type suppressNetworkEventKeyType struct{}

var suppressNetworkEventKey = suppressNetworkEventKeyType{}

// WithSuppressNetworkEvent returns a context in which the HTTP adapter
// suppresses standalone `network` event emission for outbound calls.
// Bytes are still recorded into the per-task NetworkAccountant; only
// the per-event row is withheld.
//
// LLM client wrappers (WrapOpenAI etc.) attach this to their request
// context so a single LLM call doesn't produce both an llm_call event
// AND a network event.
func WithSuppressNetworkEvent(ctx context.Context) context.Context {
	return context.WithValue(ctx, suppressNetworkEventKey, true)
}

// IsNetworkEventSuppressed reports whether the context is inside a
// WithSuppressNetworkEvent scope.
func IsNetworkEventSuppressed(ctx context.Context) bool {
	v, ok := ctx.Value(suppressNetworkEventKey).(bool)
	return ok && v
}

File: typescript/src/core/context.ts
/**
 * Task context propagation using AsyncLocalStorage.
 *
 * Allows automatic association of cost events with the currently-active
 * task without explicit parameter passing.
 */

import { AsyncLocalStorage } from "node:async_hooks";
import type { Task } from "./models.js";

const taskStore = new AsyncLocalStorage<Task>();

/**
 * Return the currently-active Task, or undefined if no task context is set.
 */
export function getCurrentTask(): Task | undefined {
  return taskStore.getStore();
}

/**
 * Execute `fn` with `task` as the current task context.
 *
 * Any code running inside `fn` (including async continuations) can call
 * `getCurrentTask()` to retrieve the task.
 */
export function runWithTask<T>(task: Task, fn: () => T): T {
  return taskStore.run(task, fn);
}

/**
 * Set the current task for the remaining async execution context.
 *
 * Uses `AsyncLocalStorage.enterWith()` (Node 18+) so the task is visible
 * to all subsequent code in the current async chain without wrapping in
 * a callback.
 */
export function setCurrentTask(task: Task): void {
  taskStore.enterWith(task);
}

// ---------------------------------------------------------------------------
// DexcostContext — dynamic customer/project attribution
// ---------------------------------------------------------------------------

/**
 * Ambient context for automatic cost attribution.
 *
 * Set once at request/job start (e.g., in Express middleware) and any
 * auto-instrumented LLM call without an explicit task will inherit
 * customerId, projectId, and metadata from this context.
 */
export interface DexcostContext {
  customerId?: string;
  projectId?: string;
  metadata?: Record<string, unknown>;
  agent?: string;
}

const contextStore = new AsyncLocalStorage<DexcostContext>();

/**
 * Set the ambient DexcostContext for the current async execution context.
 *
 * Uses `AsyncLocalStorage.enterWith()` (Node 18+) so the context is visible
 * to all code in the remaining async chain without wrapping in a callback.
 *
 * For concurrent request isolation (web servers), call this inside a
 * per-request middleware or use `runWithTask()` which isolates its own
 * async context.
 */
export function setContext(ctx: {
  customerId?: string;
  projectId?: string;
  metadata?: Record<string, unknown>;
  agent?: string;
}): void {
  contextStore.enterWith({
    customerId: ctx.customerId,
    projectId: ctx.projectId,
    metadata: ctx.metadata ?? {},
    agent: ctx.agent,
  });
}

/**
 * Return the current DexcostContext, or undefined if none is set.
 */
export function getContext(): DexcostContext | undefined {
  return contextStore.getStore();
}

/**
 * Run `fn` with `ctx` as the ambient DexcostContext, scoped to the call.
 *
 * Prefer this over `setContext()` in queue consumers, cron jobs, and any
 * long-lived worker loop: `setContext()` uses `enterWith()`, which binds the
 * context to the REMAINDER of the current async chain — consecutive jobs
 * processed on the same chain silently inherit the previous job's
 * customer/project attribution. `runWithContext()` scopes the context to
 * `fn` exactly (including all async continuations started inside it) and
 * restores the outer context afterwards.
 *
 * The context object also serves as the session grouping key, so each
 * `runWithContext()` scope gets its own ambient session task.
 */
export function runWithContext<T>(
  ctx: {
    customerId?: string;
    projectId?: string;
    metadata?: Record<string, unknown>;
    agent?: string;
  },
  fn: () => T,
): T {
  return contextStore.run(
    {
      customerId: ctx.customerId,
      projectId: ctx.projectId,
      metadata: ctx.metadata ?? {},
      agent: ctx.agent,
    },
    fn,
  );
}

/**
 * Clear the ambient DexcostContext for the current async execution context.
 */
export function clearContext(): void {
  contextStore.enterWith({});
}

// ---------------------------------------------------------------------------
// Per-call network-event suppression flag (v1 §5.3 invariant)
// ---------------------------------------------------------------------------
//
// When set, the HTTP adapter still records bytes into the NetworkAccountant
// but does NOT emit a standalone `network` event. LLM instruments wrap
// their outbound fetch call inside suppressNetworkEvent(...) so each call
// produces at most one of {llm_call, external_cost, network} — the v1
// §5.3 "≤ 1 event per HTTP call" invariant.

const suppressStore = new AsyncLocalStorage<boolean>();

/**
 * Returns `true` when the current async scope has suppressed `network`
 * event emission. Bytes are still counted into the accountant; only the
 * standalone event is withheld.
 */
export function isNetworkEventSuppressed(): boolean {
  return suppressStore.getStore() === true;
}

/**
 * Run `fn` (sync or async) inside a scope where `network`-event emission
 * is suppressed. Used by LLM instruments around their fetch call so the
 * call doesn't double-emit (llm_call + network).
 */
export function suppressNetworkEvent<T>(fn: () => T): T {
  return suppressStore.run(true, fn);
}
